### PR TITLE
Issue #10284 Ignore alternate stylesheets

### PR DESCRIPTION
**Pull request by mylainos**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10319)

### CODE REVIEW

**Comment by KiChjang** [comment]
The change to MANIFEST.json looks very wrong; did you use `./mach create-wpt /path/to/test/files` to create your test files?

**Comment by Mylainos** [comment]
Yes, I have used `./mach create-wpt --reftest tests/wpt/web-platform-tests/html/semantics/links/linktypes/alternate-css.html --reference tests/wpt/web-platform-tests/html/semantics/links/linktypes/alternate-css-ref.html`

**Comment by jdm** [comment]
The manifest thing is a known issue and won't cause any problems being merged.

**Comment by Mylainos** [comment]
*[components/script/dom/htmllinkelement.rs, line 93 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10319#-KEEA9ulAgFUVkSDzXxk:-KEEA9umSg2UUBomVzR1:-946342105) ([raw file](https://github.com/servo/servo/blob/82d0d889351a899192437dc04a0251bb10e6fafe/components/script/dom/htmllinkelement.rs#L93)):*
It may be better to do that :
```rust
let mut substrings = value.split(HTML_SPACE_CHARACTERS);
let mut substrings2 = substrings.clone();
substrings.any(|s| s.eq_ignore_ascii_case("stylesheet"))
&& !substrings2.any(|s| s.eq_ignore_ascii_case("alternate"))
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10319)*

**Comment by Mylainos** [comment]
There is a merge conflict in MANIFEST.json, what should I do ? I have error if I keep my changes or if I keep the new ones.

**Comment by jdm** [comment]
@Mylainos It's usually easiest to keep the new changes and run `./mach test-wpt --manifest-update SKIP_TESTS`.

**Comment by Mylainos** [comment]
There is a lot of errors like this one:
```
  ▶ TIMEOUT [expected OK] /2dcontext/compositing/2d.composite.canvas.destination-atop.html

  ▶ Unexpected subtest result in /2dcontext/compositing/2d.composite.canvas.destination-atop.html:
  └ NOTRUN [expected PASS] Canvas test: 2d.composite.canvas.destination-atop
```

**Comment by jdm** [comment]
Oh, that wasn't supposed to end up running any tests. Odd. Anyways, you can ignore those.

**Comment by Mylainos** [comment]
There is no error when I do `./mach test-wpt --manifest-update SKIP_TESTS` but when I run the wpt tests there is a lot of errors.

**Comment by jdm** [comment]
Debug builds in particular are known to have problems with tests in 2dcontext/, so I wouldn't worry about it too much. Running the full testsuite locally usually isn't a great use of your time; it's better to focus on particular tests or directories that are relevant.

**Comment by Mylainos** [comment]
Ok thanks.

**Comment by Mylainos** [comment]
How could I avoid this error `The command "bash etc/ci/manifest_changed.sh" exited with 1.` ?

**Comment by jdm** [comment]
That happens when the build machines run `./mach test-wpt --manifest-update SKIP_TESTS` and see the MANIFEST.json file get modified. You should include the changes to that file in this PR.

**Comment by Mylainos** [comment]
Done !

**Comment by emilio** [comment]
Can you squash the commits? They look good to me! :)

**Comment by emilio** [comment]
-S-awaiting-review 

---

Reviewed 4 of 6 files at r1, 1 of 1 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/htmllinkelement.rs, line 93 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10319#-KEEA9ulAgFUVkSDzXxk:-KESzkwJi9GkSe31k2EZ:224469717) ([raw file](https://github.com/servo/servo/blob/82d0d889351a899192437dc04a0251bb10e6fafe/components/script/dom/htmllinkelement.rs#L93)):*
It might be a bit more performant this way:

```rust
let mut found_stylesheet = false;
for s in value.split(HTML_SPACE_CHARACTERS).iter() {
    match_ignore_ascii_case! {
        "stylesheet" => found_stylesheet = true,
        "alternate" => return false, // We don't load alternate stylesheets
         _ => {},
    }
}
found_stylesheet
```

Though arguably less readable.

In practice, I don't think it matters too much, since I don't think we expect a huge amount of values in a `rel` attribute, so feel free to either make the change or leave it like it is now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10319#-:-KET-tCVFvLlYIveq_jK:879583689)*

**Comment by Mylainos** [comment]
I have this error when I test your code `error: macro undefined: 'match_ignore_ascii_case!'`

**Comment by emilio** [comment]
Yeah, sorry, it's defined in cssparser and we're on script so it's not on scope. The inner part of the loop is equivalent to:

```rust
if s.eq_ignore_ascii_case("alternate") {
    return false;
}

if s.eq_ignore_ascii_case("stylesheet") {
    found_stylesheet = true;
}
```

**Comment by Mylainos** [comment]
`.iter()` don't work so i used `.into_iter()`.

**Comment by emilio** [comment]
Looks great, thanks!

@bors-servo: r+

File: components/script/dom/htmllinkelement.rs [hunk 1]
```diff
@@ -89,8 +89,17 @@ fn get_attr(element: &Element, local_name: &Atom) -> Option<String> {
...
         Some(ref value) => {
-            value.split(HTML_SPACE_CHARACTERS)
-                .any(|s| s.eq_ignore_ascii_case("stylesheet"))
+            let mut found_stylesheet = false;
+            for s in value.split(HTML_SPACE_CHARACTERS).into_iter() {
+                if s.eq_ignore_ascii_case("alternate") == true {
+                    return false;
+                }
+
+                if s.eq_ignore_ascii_case("stylesheet") == true {
```
Comment (KiChjang) [inline components/script/dom/htmllinkelement.rs]: @bors-servo r-

nit: You can drop the `== true` here and the one above as well.

Comment (emilio) [inline components/script/dom/htmllinkelement.rs]: Whoops, missed that, @KiChjang is right

Comment (Mylainos) [inline components/script/dom/htmllinkelement.rs]: I forgot to remove it. I had a an error when I was using .iter() which said it didn't know the type.

**Comment by emilio** [comment]
Now there it goes!

@bors-servo: r+

**Pull request merged**